docs/internal/GeoffsRecoilAssessment.md

```diff
@@ -0,0 +1,176 @@
+
+I rewrote the application without MobX and using only using React's context API and useEffect to respond to state changes. The application state remained in a hierarchy, but without any observer/observable pattern.  The state was passed down the component hierarchy via context.  Container components detected state changes via additional useEffect calls. In order to coordinate larger reset of the application state (such as clearing most data), an application pub/sub mechanism was used to communicate between components to orchestrate the sequence of clearing state. This was because the useEffect order was difficult to guarantee.
+
+Finally, I rewrote the application to use Recoil. I flattened the application state into atoms. Because relationships between the data create a logical hierarchy, I wrote utility functions to make cross referencing between data islands easy. I then wrote selectors to shape the data for my React components. The container components leveraged useRecoilValue and useRecoilState to respond to state changes. Because the same application state change be updated from different React components, I implemented a dispatcher pattern based on useRecoilCallback.  I had several missteps as I learned how Recoil works, and discovered how my application was affected by flattening the state.
+
+# Recoil API: Pros, Cons, Unknowns, and Hazards
+
+## Overall
+- PRO: Built for hooks from the start and follows React concepts
+- PRO: Atom and selectors allow for composing applications from small pieces without introducing unnecessary dependencies.
+- PRO: Significantly reduces the complexity and code volume in components.
+- PRO: Looking toward the future of asynchronous, multi-threaded Javascript.
+- PRO: Recoil API surface is tiny
+- PRO: Lightning fast (at least for my application size)
+- CON: Still experimental, few releases, limited contributes - not yet mature or fully tested.
+
+## Atoms
+- PRO: Like their name, atoms let you model your data as small and independent plain types.
+- PRO: Serializing atoms to storage is straightforward compared to trying to rehydrate an observed state tree.
+- PRO: The default state of an atom with the reset() hook is convenient for clearing application state.
+- CON: Atoms are often arrays of items which means needing to look up items. This requires more undefined checks for cases where items are not found.
```

Comment:
I don't understand this objection. If an atom is defined as an array, then it should be because some natural part of our state is an array, and in that case we know already that we'll need to be smart about undefined checks. That's not part of Recoil, that's part of array handling no matter what system you use.

Reply:
I think the CON is that you have to deal with items not found in the array - vs if you can express required relationships in a state tree.

If you have a state tree where let's say an order always points at a customer and a product, you can enforce that you don't allow any order to be created with those items null.  If you have an order with a product ID and a customer ID and then have an array of order, an array of products, and and array of customers, then you will have to look them up.  If you have a bug in your state setters you may not find product ID 123 in the list of products.